### ISSUE TITLE
Clarification: Differences between switches

### ISSUE THREAD

**Issue by snowdd1** [KhaosT/HAP-NodeJS#134]
It might just be my limited knowledge of the english language:
What are differences between the three "switch" types, in layman's words?

**StatefulProgrammableSwitch**
https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js#L1669
-> "event" value of 0 or 1
-> "return" value of 0 or 1

**StatelessProgrammableSwitch**
https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js#L1686
-> "event" value of 0 or 1

**Switch**
https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js#L1702
-> Boolean

Especially the distinction between "Switch" and "StatelessProgrammableSwitch" isn't really clear to me.

**Comment by desvdp** [comment]
Have a look here: https://developer.xamarin.com/guides/ios/platform_features/introduction_to_ios9/homekit/#Working_with_Accessories

where one can read:
- StatefulProgrammableSwitch - Defines a programmable switch that remains in a give state once triggered (like a flip switch).
- /StatelessProgrammableSwitch/ - Defines a programmable switch that returns to its initial state after being triggered (like a push button)
- Switch - Defines an on/off switch like a standard wall switch

**Comment by desvdp** [comment]
Ok ... even better: have a look at the video https://developer.apple.com/videos/wwdc/2015/?id=210
in specific at 38:00

**Comment by snowdd1** [comment]
Cool, that session was really good. 
I just learned the programmable switch is more a sensor than an actuator, while the switch is an actuator.

**Comment by maddox** [comment]
> StatelessProgrammableSwitch

ohhhhh this is NICE. I just [bridged Scenes from HA](https://github.com/nfarina/homebridge/pull/177) as switches and this will be MUCH better to use.

**Comment by toshakluev** [comment]
I have a question, guys.. as i setup a stateless switch, my home app show it like an info field and not like a button, i can press it but server shows nothing... how can i setup a switch like a pushbutton with only one way, not two like "StatefullProgrammableSwitch" or just "Switch"?

**Comment by justme-1968** [comment]
you can add a timeout that will reset the state and push it to homekit.

**Comment by toshakluev** [comment]
Ahhh... Its a boring solution :)) i need apple style, just a push button, not less, not more

**Comment by dustindclark** [comment]
@antonkluev, did you find a solution on this?  I don't get inside of my setState function with the code below but similar code works with a Switch.  Maybe I'm missing something?

```
this.service = new Service.StatelessProgrammableSwitch(this.name);
this.service
                .addCharacteristic(Characteristic.On)
                .on('set', this.setState.bind(this));
```

**Comment by antonkluev** [comment]
no, i think there is a bug )) we must ask the creator of HAP

**Comment by dannyhertz** [comment]
Anyone know if the StatelessProgrammableSwitch service is working yet? I can't seem to get it to trigger events either and I'd like to use this for firing IR signals to turn on an AC (which just uses a power toggle vs on/off state)

**Comment by KhaosT** [comment]
It works fine. Please make sure you properly invoked the setValue so iOS can receive the value changes.

**Comment by dannyhertz** [comment]
@KhaosT hmm sorry for being a bit new to the Homebridge system, but I can't seem to get a simple StatelessProgrammableSwitch working. Could you let me know if something here looks off?

This is the basic structure of my accessory:

```
LircAirConditioner.prototype = {

  sendPowerCommand: function(value, callback) {
    console.log("Sending power command: ", this.lirc_name, this.power_command);
    lircClient.irsend.send_once(this.lirc_name, this.power_command, function() {
      console.log("Sent AC power command!");
      callback(null, "off");
    });
  },

  getServices: function() {
    var informationService = new Service.AccessoryInformation();
    informationService
      .setCharacteristic(Characteristic.Manufacturer, this.manufacturer)
      .setCharacteristic(Characteristic.Model, this.model_name)

    var powerSwitchService = new Service.StatelessProgrammableSwitch(this.name + " Power");
    powerSwitchService
      .getCharacteristic(Characteristic.ProgrammableSwitchEvent)
      .on('set', this.sendPowerCommand.bind(this));

    return [informationService, powerSwitchService];
  }
  
}
```

The lircClient is a library that sends IR signals on my rasberry Pi, but that shouldn't really make a difference for this example. Currently I can see the accessory in the Home app on iOS but is says its not compatible. 

![img_0017](https://cloud.githubusercontent.com/assets/407719/18761404/c5562cb8-80d2-11e6-8f0a-5bf45461ae6f.PNG)

Any help is appreciated!

**Comment by KhaosT** [comment]
Wait... what are you trying to do... Programmable Switch is used to refer to a physical switch in real world that user can press and emit event to HomeKit. It's not the other way around...
Based on your code, it looks like you are expecting HomeKit to write to that characteristic which is not possible since it's defined as read-only.

**Comment by dannyhertz** [comment]
Yeah, I'm effectively using an IR transmitter to act like a stateless button or "switch" on a remote control. There is no way of figuring out if the AC is on/off so just sending this command will toggle it. For that reason a stateless switch seemed like the right thing, but again I'm very new to this HomeKit stuff. Is that not the correct way to think about it? How could I accomplish something like that.

**Comment by PatchworkBoy** [comment]
With a regular on/off switch, which defaults to off, and uses a JS setTimeout to reset it back to off again after an on is triggered. S’how I’ve got push-to-toggle switches defined for exactly the same purpose - sending LIRc commands.

**Comment by dannyhertz** [comment]
@PatchworkBoy ah, clever! Would you mind sharing your code to accomplish that? I'm familiar with working with timeouts, but I'm more curious how you configured the characteristics and callbacks. Thanks!

**Comment by PatchworkBoy** [comment]
Ah - Don’t think my code’ll work for you I’m afraid... I’m using homebridge via platform plugins as opposed to writing natively for HAP-NodeJS. My functionality deals with the platform (Domoticz) rather than homekit, and homebridge picks this up this change via MQTT.

(Just realised I’ve never actually worked DIRECT with HAP-NodeJS!)

**Comment by rooi** [comment]
I've included a StatelessProgrammableSwitch in my plugin which triggers after pushing a volume up/down button. It seems to work, but when I hook up an automation trigger that should act after push the volume up/down button noting happens.

Does automation work with StatelessProgrammableSwitch s?

Thanks!

**Comment by kim3er** [comment]
@rooi According to @KhaosT, the HomeKit app no longer supports programable switches. Which sucks. I've tried all the free third party apps, to no avail. Not sure I'm ready to sink £11 into the third party Home app. https://github.com/KhaosT/homebridge-amazondash/commit/d8e16a399edd08c1e508428abd03901de8886bb7

**Comment by KhaosT** [comment]
@kim3er there are free options like Elgato Eve's app or Hesperus if you don't feel like spending money for home automation apps.

**Comment by kim3er** [comment]
You're right, @KhaosT! I'd completely missed the option in the Eve app. Rich

**Comment by AviFix** [comment]
@justme-1968 Can you please sho an example on how to push back the value to homeKit?

**Comment by ebaauw** [comment]
The built-in Home app in iOS 10.3 _does_ support the StatelessProgrammableSwitch service, albeit somewhat differently from what I would expect.  Looking at Apple's first public beta, I find:
- The StatelessProgrammableSwitch service seems to refer to a button on a switch device, rather than to the device itself.  A switch device with multiple buttons is a single HomeKit accessory with a different StatelessProgrammableSwitch service for each button.  I managed to add a BatteryService service to the accessory as well;
- The iOS 10.3 Home app shows a single tile for the accessory (!), with a new symbol, where the iOS 10.2 Home app shows a tile per (unsupported) StatelessProgrammableSwitch service.
- In Details, the iOS 10.3 Home app shows a section STATUS (for the BatteryService) and a section for each button (one for each StatelessProgrammableSwitch service), BUTTON 1, BUTTON 2, etc.  It ignores the service's Name.  When the accessory contains only a single StatelessProgrammableSwitch service, the section is simply labeled ACTIONS;
- Each button section has two related actions, Single Press and Double Press.  On Add you can specify the scenes and/or services to activate, very much like specifying actions in Automation.
- These actions do not show in Automation in the iOS 10.3, but they do in the iOS 10.2 and in other HomeKit apps, like Eve.  The rules for Single Press trigger on ProgrammableSwitchEvent = 0, the rule for Double Press checks for ProgrammableSwitchEvent = 1.

The last point leads to my question: how would I set the ProgrammableSwitchEvent?  Will it trigger an action each time I `buttonService.setCharacteristic(Characteristic.ProgrammableSwitchEvent, 0)` or only if the ProgrammableSwitchEvent actually changes value?  In the latter case, to what value would I set the characteristic after the timeout to reset the switch?  I don't think HomeKit appreciates an out of bound value (minValue and maxValue are 0 and 1 respectively).

**Comment by KhaosT** [comment]
@ebaauw nice to see you do the digging there :) It's kinda annoying that Home app uses a different definition of `StatelessProgrammableSwitch` service than it used to be in HAS. I think what Home app expects is that accessory will only send event when the button is being pressed ignoring the characteristic's last value. (Basically should match your first case). However, HAP-NodeJS currently implemented in a way that notification will not be send to iOS if the updated characteristic value is the same as the old one so by invoking setCharacteristic multiple times will only send one notification.

I'll try to wait and see how HAS being updated for StatelessProgrammableSwitch service and make change accordingly to the project.

**Comment by ebaauw** [comment]
> nice to see you do the digging there

@KhaosT  I really appreciate the work you've been doing with HAP-NodeJS and I'm happy to provide what little help I can.

> I'll try to wait and see how HAS being updated

HAS = HomeKit Accessory Simulator?  I don't think Apple will change it, at least not for a minor iOS release.  I reverse engineered the above using HAS Version 2.0 (114.2), before starting to implement the changes in my homebridge-hue plugin.  Also, the iOS 10.3 and iOS 10.2 Home apps (and other HomeKit apps on iOS 10.3 beta as well as on iOS 10.2) seem to interoperate without any issues.  I think the changes are limited to the builtin Home app.

It might be prudent to wait until the final release of iOS 10.3, though.

**Comment by KhaosT** [comment]
From what I heard there will have an updated version with adjusted UI for programmable switches, not sure when it will be released though.

**Comment by ebaauw** [comment]
> there will have an updated version with adjusted UI for programmable switches

Cool, looking forward to see it.

> However, HAP-NodeJS currently implemented in a way that notification will not be send to iOS if the updated characteristic value is the same as the old one so by invoking setCharacteristic multiple times will only send one notification.

Looking at the `setValue()` code, I could probably work around that by calling from my plugin the following:
```
buttonService.getCharacteristic(Characteristic.ProgrammableSwitchEvent)
  .emit('change', {oldValue: oldValue, newValue: newValue});
```

> The rules for Single Press trigger on ProgrammableSwitchEvent = 0, the rule for Double Press checks for ProgrammableSwitchEvent = 1.

Some more experimentation yields:
- When you change the ProgrammableSwitchEvent's maxValue property to 0, Home only shows the Single Press action.
- When you change the ProgrammableSwitchEvent's maxValue property to 2, Home shows a third action, Long Press. 
- When you increase the ProgrammableSwitchEvent's maxValue property above 2, Home shows additional actions, labelled (Unknown).

Since the Hue bridge lacks push notifications, I need to poll it.  Therefore I cannot detect Double Press reliably, so I won't support that action.  For the Hue tap switch, I want to support only Single Press - that's easy.  For the Hue dimmer switch I want to support Single Press and Long Press, as the dimmer supports Press and Hold, which I can detect reliably.  So I need to specify allowed values 0 and 2.  Does HAP-NodeJS support that?

**Comment by KhaosT** [comment]
@ebaauw I'm not sure you emitting event externally on behalf of the characteristic object is a good idea since it's a implementation detail which may be changed in the future. It's fine for trying it out, just don't ship it :)

Inside props, there is a [`validValues`](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Characteristic.js#L266) you can set. Try set that to an array of [0, 2] and see if Home app will honor that I guess.

**Comment by ebaauw** [comment]
I understand.  Will try out later and let you know if it works.

Bingo.  Only Single Press and Long Press after:
```
buttonService.getCharacteristic(Characteristic.ProgrammableSwitchEvent)
  .setProps({minValue: 0, maxValue: 2, validValues: [0, 2]});
```

**Comment by ebaauw** [comment]
> I'm not sure you emitting event externally on behalf of the characteristic object is a good idea

HomeKit seems OK with initialising/resetting the `ProgrammableSwichEvent` to -1, so I don't need to mess with the implementation details.

**Comment by ebaauw** [comment]
For those interested, I published some screenshots of a programmable switch in the iOS 10.3 beta 2 Home app.  The switch is for a Hue dimmer switch exposed by homebridge-hue v0.2.2.  See https://github.com/ebaauw/homebridge-hue/issues/54#issuecomment-280133303.

**Comment by ebaauw** [comment]
Now here's an interesting fact: the `ProgrammableSwitchEvent` characteristic appears neither writeable nor readable to HomeKit apps.  It's permissions are limited to notify, despite the setting in homebridge (or, for that matter, in the HomeKit Accessory Simulator) that includes read and notify permissions.

I found out playing with `HMCatalog` and noticing that the `InputEvent` wasn't marked _Read Only_.

This probably explains why Eve doesn't show this characteristic, allows it as a trigger for a rule, but not as a condition.
